module/Database/src/Service/Payment.php

```diff
@@ -0,0 +1,356 @@
+            // We have a problem, a prospective member must have at least one checkout session associated with it.
+            // Ignore it for now and create a new Checkout Session.
+            return $this->getCheckoutLink($prospectiveMember);
+        }
+
+        // We have at least one known Checkout Session on file.
+        if (
+            CheckoutSessionStates::Paid === $lastCheckoutStub->getState()
+            || CheckoutSessionStates::Pending === $lastCheckoutStub->getState()
+        ) {
+            // Checkout Session is finalised or will be after payment processing. Do not allow the prospective member to
+            // do something else.
+            return null;
+        }
+
+        if (CheckoutSessionStates::Failed === $lastCheckoutStub->getState()) {
+            // Last payment failed, so we need to create a new Checkout Session for the user to be able to try again.
+            return $this->getCheckoutLink($prospectiveMember);
+        }
+
+        if (CheckoutSessionStates::Expired === $lastCheckoutStub->getState()) {
+            // The Checkout Session has already been abandoned.
+
+            if ((new DateTime())->add(new DateInterval('PT5M')) >= $lastCheckoutStub->getExpiration()) {
+                // The Checkout Session is completely abandoned, as the maximum expiration for the recovery URL of 30
+                // days has passed (or will pass in 5 minutes). As such, we want to create a new Checkout Session for
+                // this prospective member.
+                return $this->getCheckoutLink($prospectiveMember);
+            }
+
+            // The Checkout Session is not completely dead yet, so return the recovery URL.
+            return $lastCheckoutStub->getRecoveryUrl();
+        }
+
+        // Checkout Session is still valid (at this point, not necessarily when the prospective member finally submits).
+        // We try to retrieve the actual Checkout Session, if this succeeds we return the URL. If not we fail and let
+        // the prospective member know.
+        return $this->getCheckoutSession($lastCheckoutStub->getCheckoutId())?->url;
+    }
+
+    /**
+     * Create a Checkout Session through the Stripe API. The Checkout Session has some required parameters, for more
+     * information on their details see {@link https://stripe.com/docs/api/checkout/sessions/object}.
+     */
+    private function createCheckoutSession(ProspectiveMemberModel $prospectiveMember): ?CheckoutSession
+    {
+        try {
+            return $this->getClient()->checkout->sessions->create([
+                'line_items' => [
+                    [
+                        'price' => $this->config['membership_price_id'],
+                        'quantity' => 1,
+                    ],
+                ],
+                'mode' => 'payment',
+                'cancel_url' => $this->config['cancel_url'],
+                'success_url' => $this->config['success_url'],
+                'after_expiration' => [
```

Comment:
We could expire the payment after 1 hour to send an email about failure earlier; if we then allow failed payments to be bypassed by the secretary, we can have an alternative payment method for those who really are unable to pay online (such as debit card or cash)

```suggestion
                'expires_at' => time() + 3600,
                'after_expiration' => [
```

Reply:
25b64bd